tss-esapi/src/traits.rs

```diff
@@ -51,65 +38,151 @@ pub trait UnMarshall: Sized {
+    };
+}
 
-    fn marshall_offset(
-        &self,
-        marshalled_data: &mut [u8],
-        offset: &mut std::os::raw::c_ulong,
-    ) -> Result<()> {
-        ReturnCode::ensure_success(
-            unsafe {
-                crate::tss2_esys::Tss2_MU_UINT32_Marshal(
-                    *self,
-                    marshalled_data.as_mut_ptr(),
-                    marshalled_data.len().try_into().map_err(|e| {
-                        error!("Failed to convert size of buffer to TSS size_t type: {}", e);
-                        Error::local_error(WrapperErrorKind::InvalidParam)
-                    })?,
-                    offset,
-                )
-            },
-            |ret| {
-                error!("Failed to marshall u32: {}", ret);
-            },
-        )?;
+/// A macro for implementing the Unmarshall trait
+/// for a specific TSS type.
+macro_rules! impl_unmarshall_trait {
+    ($native_type:ident, $tss_type:ident, $tss_mu_type:ident, $convert_expression:expr) => {
+        paste::item! {
+            impl $crate::traits::UnMarshall for $native_type {
+                fn unmarshall_offset(marshalled_data: &[u8], offset: &mut usize) -> Result<Self> {
+                    let mut dest = $tss_type::default();
+                    let ffi_buffer_size = $crate::ffi::FfiSizeType::try_from(marshalled_data.len())?;
+                    let mut ffi_offset = $crate::ffi::FfiSizeType::try_from(*offset)?;
+                    crate::ReturnCode::ensure_success(
+                        unsafe {
+                            crate::tss2_esys::[ < Tss2_MU_ $tss_mu_type _Unmarshal >](
+                                marshalled_data.as_ptr(),
+                                ffi_buffer_size.into(),
+                                ffi_offset.as_mut_ptr(),
+                                &mut dest,
+                            )
+                        },
+                        |ret| log::error!("Failed to unmarshal {}: {}", std::stringify!($native_type), ret),
+                    )?;
+                    *offset = usize::try_from(ffi_offset)?;
+                    $convert_expression(dest)
+                }
+            }
+        }
+    };
+}
 
-        Ok(())
-    }
+/// Macro used to implement Marshall and Unmarshall for types
+/// that are just a type aliases of native types and are passed
+/// to MUAPI by value.
+macro_rules! impl_mu_aliases {
+    ($tss_type:ident) => {
+        $crate::traits::impl_marshall_trait!($tss_type, $tss_type, $tss_type, into(),);
+        $crate::traits::impl_unmarshall_trait!($tss_type, $tss_type, $tss_type, Ok);
+    };
 }
 
-impl UnMarshall for u32 {
-    fn unmarshall_offset(
-        marshalled_data: &[u8],
-        offset: &mut std::os::raw::c_ulong,
-    ) -> Result<Self> {
-        let mut dest = 0_u32;
+/// Macro used to implement Marshall and Unmarshall for types that
+/// can be converted from native to TSS i.e. it cannot fail and are
+/// passed to MUAPI by value(i.e. pointer).
```

Comment:
`by value (i.e. pointer)` - is this correct?

Reply:
Fixed